### ISSUE TITLE
Idea: add interface upscaling 

### ISSUE THREAD

**Issue by thedarklordkitten** [raceintospace/raceintospace#359]
Dear all, I've been silently following your project for a few years. Recently I downloaded the latest version 1.1 (for linux). My laptop has a 4k screen,  and as a consequence the interface window of the game is tiny. I tried activating fullscreen by modifying the corresponding parameter in the config file,  but this only centered the interface in the middle of the screen. I wonder if you already considered using some upscaling library like those discussed here [pixel art scaling algorithms](https://en.m.wikipedia.org/wiki/Pixel-art_scaling_algorithms) . Anyways thanks for your time and effort.

**Comment by peyre** [comment]
Hi darklord!  Glad to have you checking out the game.  I'm also running Linux (specifically Xubuntu); I've never tried fullscreen before since the size of the game window has never been a problem for me.

So I turned on fullscreen in Advanced Preferences.  RIS 1.1 opens like I would expect on my system: the game runs fullscreen in my default display, and it blacks out the other monitor.   However, when I opened RIS 1.2, the game ran normally.  I don't know what was going wrong with RIS 1.1 on your system, but it looks like the fullscreen option just doesn't work in RIS 1.2.  I'll change the name of this Issue accordingly.

**Comment by thedarklordkitten** [comment]
Thanks for the quick reply. I got back to my laptop and grabbed a pair of pictures, so that the issue is easier to point out. The RIS version I refer to is the latest build found on sourceforge, installed on linux Kubuntu 20.04.
The first screenshot is the capture of the windowed mode: as you see in my 4k monitor it is really a small window.
![screenshot_ris_4k](https://user-images.githubusercontent.com/66200568/83387291-c327db80-a3ec-11ea-8127-fa6040f2fa4e.png)
The second image (sorry for the quality) is a photo of my laptop's screen (the screenshot utility wouldn't come up) when the game is started with the option "fullscreen 1" set in the config file.
![picture_ris_fullscreen_4k](https://user-images.githubusercontent.com/66200568/83387308-c9b65300-a3ec-11ea-8b85-810fb3fe9bd4.png)
Hope this helps clarifying the issue, let me know if I can provide you with more info.

**Comment by peyre** [comment]
Strange, I've never seen it do that!  Still, that's the old Sourceforge RIS and I don't think we can do much with that; it hasn't been maintained since 2013.  Am I right in thinking you're running Kubuntu?  If so, you can download and compile the most recent version of the game.  The instructions at https://github.com/raceintospace/raceintospace are step-by-step and work reliably (I use them myself).

**Comment by thedarklordkitten** [comment]
Ok, I manage to set up a build environment in Kubuntu 20.04.
If I run it in windowed mode, since the monitor is the same, I still get the tiny window.

![ris_latest_04june2020_4k](https://user-images.githubusercontent.com/66200568/83728857-1d23dd80-a647-11ea-8039-0da35a9edbdf.png)

This was run using "make run" command. Where is the config file located? I couldn't manage to find one and try fullscreen.
Thanks :)

**Comment by peyre** [comment]
Yeah, I guess that's just how it shows when your resolution is crazy high.

The config file (it has no extension) is located in ~/.raceintospace.  You can open it with a text editor such as mousepad.

**Comment by peyre** [comment]
darklord, were you able to get it working right?  If so I'd like to close the ticket; if you're having trouble finding ~/.raceintospace,  that's /home/%username%/.raceintospace (note the dot), e.g. /home/kitten/.raceintospace.  You probably know all that already; I just don't want to make any assumptions about your level of technical knowhow.

**Comment by thedarklordkitten** [comment]
Sorry for the delay, I had a busy few days. I can confirm I cannot manage to get v1.2 fullscreen as you pointed out earlier. I will try to upload another screenshot as soon as possible, but I think there is nothing else to report about this issue. Thanks for your time :)

**Comment by laenion** [comment]
With v1.2 the problem is that the fullscreen setting is simply ignored :-) The desired value is passed to https://github.com/raceintospace/raceintospace/blob/master/src/display/graphics.cpp#L39, but then simply ignored. I'll submit a pull request in a few minutes to fix this, passing `SDL_FULLSCREEN` to `SDL_SetVideoMode` again.

However this won't change the original problem with the tiny window: By default the original 320x200 resolution will be scaled by factor 2. As documented in https://www.libsdl.org/release/SDL-1.2.15/docs/html/sdlsetvideomode.html SDL1 will try to set the next available higher resolution, which *should* be 640x480, but if the system / graphics card / driver doesn't support that, this may result in that tiny fullscreen image from above.

One possibility may be to add a command line option to let the user specify the scaling factor. Or the game could try to calculate a better scaling factor itself. Or everything could be ported to SDL2 to be able to use `SDL_WINDOW_FULLSCREEN_DESKTOP`, which will do the scaling in software to work around limited hardware support...

**Comment by peyre** [comment]
Sounds like great work there (though I'm not a programmer myself so can't really judge).  Would you be interested in joining the team?  We could probably use some help with the 2.0 release, among other things.

**Comment by laenion** [comment]
I'd love to, but I fear I simply won't have the time to do so - my backlog is quite crowded already.
You *may* try to assign me a ticket and see what happens though ;-)

Just be aware that my knowledge is limited to Linux systems, and I don't know the game at all - I just wanted to get it working fullscreen because I love huge pixels :-)

**Comment by peyre** [comment]
If it matters, I've tried it myself and the fullscreen option doesn't seem to work at all.

**Comment by peyre** [comment]
I think it's fixed now.  Would you check it and let me know whether to close this Issue?

**Comment by Alsed** [comment]
> Dear all, I've been silently following your project for a few years. Recently I downloaded the latest version 1.1 (for linux). My laptop has a 4k screen, and as a consequence the interface window of the game is tiny. I tried activating fullscreen by modifying the corresponding parameter in the config file, but this only centered the interface in the middle of the screen. I wonder if you already considered using some upscaling library like those discussed here [pixel art scaling algorithms](https://en.m.wikipedia.org/wiki/Pixel-art_scaling_algorithms) . Anyways thanks for your time and effort.

Thanks for commenting on these filters. I have managed to get the game to scale to 4x using the HQ4x algorithm.
I had previously had the internal Scale2x algorithm attempt higher scaling, but the results were notoriously bad. 
Eventually we could set filters for 3x and 4x or more, and set in the configuration file the option to choose it.

![Captura desde 2024-06-07 14-32-57](https://github.com/raceintospace/raceintospace/assets/7539842/aa5d4f97-2c00-482b-b12e-e7bf98eabbc9)
Scaled with HQ4x

**Comment by peyre** [comment]
That looks great, Alsed!  Can we implement this?

**Comment by Alsed** [comment]
Yes, it's fully functional. I have to upload the changes to the repository and pull them.
I think it's best to leave this resolution (4x) by default, but I could spend some time configuring so that we can choose the scaling, and even implement a 3x algorithm. Although I don't know if it's really necessary.

**Comment by peyre** [comment]
I would like very much if people could change the resolution in the config file.  Many (possibly myself included) might prefer the original size.

**Comment by Alsed** [comment]
Hi @Peyre I've set the changes so it can be configure in the advanced options file.
Default now is 4x, but you can change to classic 2x.
https://github.com/raceintospace/raceintospace/pull/852

**Comment by peyre** [comment]
Hi Alsed!  I just compiled the latest code and tried it out.  It does indeed default to 4X.  However, setting **4xscale 0** in the config file doesn't change it; it still shows in 4X.  I use small monitors so 2X looks best on it.

I wonder, would it be doable for users to make it even bigger, like if someone's using a huge 32" monitor or something?  Like maybe **4xscale** shouldn't be grayed out, but it would be set to 1 by default, 0 would set 2X, 3 would set 6X, etc.

**Comment by Alsed** [comment]
Hi Peyre, i would look into it about the 2x option not working. 

There is a world of possibilities with Image Scaling, even with simple techniques there are a ton of them.
Advanced techniques use opengl shaders and even machine learning.
https://en.wikipedia.org/wiki/Image_scaling

Scaling and fullscreen options work independently. From what i've tested, there is no noticeable improvement in image quality either using 2x fullscale or 4x fullscale.

**Comment by peyre** [comment]
No, the image quality doesn't change, but my screens are only 17" so 4X RIS bleeds over onto the right screen, and the whole thing looks noticeably more pixelated.

That makes sense, I think, that fullscreen would work differently.  I see it overrides the 4xscale setting; I might add that as a comment to the config file later.

**Comment by peyre** [comment]
I downloaded and compiled after merging the PR, but the option in config still doesn't do anything.

**Comment by Alsed** [comment]
I'm still figuring out why editing the config file doesn't seem to work. 
Meanwhile, you can run in the terminal 'raceintospace -s' (without the ') and the game will run at 2x scale.

To try this out, I installed the game with 'make install' but it seems that cmake doesn't have an 'uninstall' command.

**Comment by peyre** [comment]
Yes, that works with the -s option.  That may be a workaround if it comes down to it.
